clients/auth.go

```diff
@@ -0,0 +1,257 @@
+}
+
+func (b *authClientBuilder) WithHostGetter(hostGetter disc.HostGetter) *authClientBuilder {
+	b.hostGetter = hostGetter
+	return b
+}
+
+func (b *authClientBuilder) WithTokenProvider(tokenProvider TokenProvider) *authClientBuilder {
+	b.tokenProvider = tokenProvider
+	return b
+}
+
+func (b *authClientBuilder) Build() *AuthClient {
+	if b.hostGetter == nil {
+		panic("authClient requires a hostGetter to be set")
+	}
+	if b.tokenProvider == nil {
+		panic("authClient requires a tokenProvider to be set")
+	}
+
+	if b.httpClient == nil {
+		b.httpClient = http.DefaultClient
+	}
+
+	return &AuthClient{
+		httpClient:    b.httpClient,
+		hostGetter:    b.hostGetter,
+		tokenProvider: b.tokenProvider,
+	}
+}
+
+// ListUserRestrictedTokens listsrestricted tokens for a given user
+func (client *AuthClient) ListUserRestrictedTokens(userID string, token string) (RestrictedTokens, error) {
+	host := client.getHost()
+	if host == nil {
+		return nil, errors.New("No known auth hosts")
+	}
+	host.Path = path.Join(host.Path, "v1", "users", userID, "restricted_tokens")
+
+	req, _ := http.NewRequest("GET", host.String(), nil)
+	req.Header.Add("x-tidepool-session-token", token)
+
+	res, err := client.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if res.StatusCode == 200 {
+		retVal := RestrictedTokens{}
+		if err := json.NewDecoder(res.Body).Decode(&retVal); err != nil {
+			log.Println(err)
+			return nil, &status.StatusError{
+				Status: status.NewStatusf(res.StatusCode, "ListUserRestrictedTokens Unable to parse response[%s]", req.URL),
+			}
+		}
+		return retVal, nil
+	} else if res.StatusCode == 404 {
+		return nil, nil
+	} else {
+		return nil, &status.StatusError{
+			Status: status.NewStatusf(res.StatusCode, "Unexpected response code from service[%s]", req.URL),
+		}
+	}
+}
+
+// CreateRestrictedToken creates a restricted token for a given user
+func (client *AuthClient) CreateRestrictedToken(userID string, expirationTime time.Time, paths []string, token string) (*RestrictedToken, error) {
+	host := client.getHost()
+	if host == nil {
+		return nil, errors.New("No known auth hosts")
+	}
+	host.Path = path.Join(host.Path, "v1", "users", userID, "restricted_tokens")
+
+	payload := RestrictedTokenCreate{
+		Paths:          &paths,
+		ExpirationTime: &expirationTime,
+	}
+
+	if jsonToken, err := json.Marshal(payload); err != nil {
+		return nil, fmt.Errorf("unable to marshal payload: %w", err)
+	} else {
+		req, _ := http.NewRequest("POST", host.String(), bytes.NewBuffer(jsonToken))
+		req.Header.Add("x-tidepool-session-token", token)
+
+		res, err := client.httpClient.Do(req)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't create restricted token")
+		}
+		defer res.Body.Close()
+
+		switch res.StatusCode {
+		case http.StatusCreated:
+			var td RestrictedToken
+			if err = json.NewDecoder(res.Body).Decode(&td); err != nil {
+				log.Println("Error parsing JSON results", err)
+				return nil, nil
```

Comment:
I think you should return the error here instead of just logging and ignoring it.